K8s auto-instrumentation details and example (Java)

eedugon opened the pull request:
This PR adds the first `language-oriented` document for instrumenting applications on Kubernetes.

The idea is to create similar docs for all supported languages.

@akhileshpok and @mlunadia : I'd like your review on this before moving forward to other languages.

I think this is a good learning resource as we show the details of how auto-instrumentation is achieved (each language works in a different way).

I'll also ping the owners of the SDKs to see if they want to include some extra details about the specific language here.

cc: @gizas , @rogercoll , @bmorelli25 and @mdbirnstiehl

jackshirazi commented on docs/kubernetes/operator/instrumenting-java.md:
````diff
@@ -0,0 +1,183 @@
+elastic-instrumentation   107s   http://opentelemetry-kube-stack-daemon-collector.opentelemetry-operator-system.svc.cluster.local:4318
+```
+
+Auto-instrumentation steps:
````
```suggestion
Example auto-instrumentation steps:
```

jackshirazi commented on docs/kubernetes/operator/instrumenting-java.md:
```diff
@@ -0,0 +1,183 @@
+  Ensure the environment variable `OTEL_EXPORTER_OTLP_ENDPOINT` points to a valid endpoint and there's network communication between the Pod and the endpoint.
+
+4. Confirm data is flowing through in **Kibana**:
+<!--> Should we add some screenshots here? <-->
```
screenshots get out of date quickly

jackshirazi commented on docs/kubernetes/operator/instrumenting-java.md:
```diff
@@ -0,0 +1,183 @@
+
+  Ensure the environment variable `OTEL_EXPORTER_OTLP_ENDPOINT` points to a valid endpoint and there's network communication between the Pod and the endpoint.
+
+4. Confirm data is flowing through in **Kibana**:
```
There can be an intermediate step which is to confirm that the endpoint (the collector or apm server) is getting the data, but that's more detailed than might be worth putting in here

eedugon replied:
@jackshirazi , either here or in the dedicated `troubleshooting` doc, that will definitely be important. Thanks for reminding it.

how would you verify that exactly? Are there specific logs in the DaemonSet collector (the current endpoint) about the reception and export of traces? or would you look into the application logs for that? I haven't dig yet into the logs details, but it's something we will have to address soon.

jackshirazi replied:
I haven't checked yet. I think you can set the collector to output the requests to log/console, that's probably how I'd check

eedugon replied:
That approach should better fit in a troubleshooting doc than here (making note of it for the next troubleshooting doc update). In case there's a straight forward way to determine that traces are flowing with the default installation configuration I'll add it at a later stage.
Thanks!

eedugon commented:
@jackshirazi , I have removed the provided example after a conversation with @mlunadia .
Adding an example with a real container image would force us to ensure its maintained properly, so lets keep this a bit simpler.

Would you please take a look and share your thoughts to the updated content?

eedugon commented:
@mdbirnstiehl : I've added you as reviewer here because we are about to merge this PR now that the content has been validated and agreed.

@jackshirazi , I finally added two small changes if you want to take a look:
- I added a link to the upstream auto-instrumentation -> java doc at the beginning of the doc.

- I added in the main `instrumenting_applications.md` a section `how auto-instrumentation works` where we explain in general how the operator injects and configures the instrumentation SDK, and it's now referenced in the Java related file. I think that adds value in the generic doc (language agnostic).

mdbirnstiehl commented on docs/kubernetes/operator/instrumenting-java.md:
```diff
@@ -0,0 +1,154 @@
+    - The application appears in the list of services.
+    - The application shows transactions and metrics.
+  
+  - For application container logs, open **Kibana Discover** and filter for your pods log. In the provided example we could filter them with any of:
```
```suggestion
  - For application container logs, open **Kibana Discover** and filter for your Pod's logs. In the provided example, we could filter for them with either of the following:
```
Not sure if this should be Pods' or Pod's. Is it one Pod or multiple?

mdbirnstiehl commented on docs/kubernetes/operator/instrumenting-java.md:
```diff
@@ -0,0 +1,154 @@
+
+4. Verify the [auto-instrumentation resources](./instrumenting-applications.md#how-auto-instrumentation-works) are injected in the Pod:
+
+  Run a `kubectl describe` of one of your application pods and check:
```
```suggestion
  Run a `kubectl describe` of one of your application Pods and check:
```

mdbirnstiehl commented on docs/kubernetes/operator/instrumenting-java.md:
```diff
@@ -0,0 +1,154 @@
+
+## Java agent extensions consideration
+
+The operator supports configuration that installs [Java agent extensions](https://opentelemetry.io/docs/zero-code/java/agent/extensions/) in `Instrumentation` objects. The extension needs to be available in an image. Refer to [using extensions with the OpenTelemetry Java agent](https://www.elastic.co/observability-labs/blog/using-the-otel-operator-for-injecting-elastic-agents#using-an-extension-w ...
```
```suggestion
The operator supports a configuration that installs [Java agent extensions](https://opentelemetry.io/docs/zero-code/java/agent/extensions/) in `Instrumentation` objects. The extension needs to be available in an image. Refer to [using extensions with the OpenTelemetry Java agent](https://www.elastic.co/observability-labs/blog/using-the-otel-operator-for-injecting-elastic-agents#using-an-extension-with-the-opentelemetry-java-agent) for an example of adding an extension to an agent.
```

mdbirnstiehl reviewed (commented):
Overall, looks good! Just added a few comments mostly grammar suggestions.

eedugon commented on docs/kubernetes/operator/instrumenting-java.md:
```diff
@@ -0,0 +1,154 @@
+
+3. Restart application
+
+  Once the annotation has been set, the Pods need to be recreated for the instrumentation libraries to be injected.
```
I'm going to change it to something like "`restart the application` to force the recreation of the Pods" or something like that, as there's no such action like `recreate` the Pods.

The idea is that either the user deletes Pods (the workload controller creates new Pods) or restart the application through `kubectl rollout restart`.

jackshirazi commented on docs/kubernetes/operator/instrumenting-applications.md:
```diff
@@ -152,6 +159,32 @@ kubectl annotate namespace java-apps instrumentation.opentelemetry.io/inject-jav
+- **Configuring the main container**:
+
+  The operator injects environment variables into the main application container to configure OpenTelemetry settings (for example, `OTEL_EXPORTER_OTLP_ENDPOINT` or `OTEL_TRACES_SAMPLER`). Additionally, it links the instrumentation library to the application using mechanisms specific to the language runtime, such as:
+    - **For Java**: The library is linked through the `javaagent` option.
```
```suggestion
    - **For Java**: The library is linked through the `javaagent` option using the JAVA_TOOL_OPTIONS environment variable.
```
